### PR TITLE
Change/better parsing

### PR DESCRIPTION
**Pull request by shaharyarn**

### CODE REVIEW

File: marine.c [hunk 1]
```diff
@@ -281,46 +293,38 @@ marine_write_specified_fields(packet_filter *filter, epan_dissect_t *edt, char *
...
-            }
+            output[counter] = (gchar *) g_malloc0(get_field_length(fv_p));
+            int field_counter = 0;
 
             /* Output the array of (partial) field values */
             for (j = 0; j < g_ptr_array_len(fv_p); j++) {
                 str = (gchar *) g_ptr_array_index(fv_p, j);
                 for (char *p = str; *p != '\0'; p++) {
-                    output[counter++] = *p;
+                    output[counter][field_counter++] = *p;
```
Comment (yehonatanz) [inline marine.c:316]: Why not `strncpy`?

File: marine.c [hunk 2]
```diff
@@ -520,17 +523,12 @@ WS_DLL_PUBLIC marine_result *marine_dissect_packet(int filter_id, unsigned char
     } else {
         int *key = packet_filter_keys[filter_id];
         packet_filter *filter = (packet_filter *) g_hash_table_lookup(packet_filters, key);
-        char *output = filter->output_fields == NULL ? NULL : (char *) g_malloc0(4096); // TODO export to const
+        unsigned int output_count = filter->expected_output_len;
+        char **output = filter->output_fields == NULL ? NULL : (char **) g_malloc0(sizeof(char *) * output_count); // TODO export to const
```
Comment (yehonatanz) [inline marine.c]: Remove the `TODO` :)

**Review by yehonatanz — COMMENTED** [review]
Looks really good!

File: marine.c (hunk 1 above)
Comment (domerd) [reply marine.c:316]: `strncpy` adds padding in case a NULL character is reached before the given string.

**Review by yehonatanz — APPROVED** [review]

File: marine.c [hunk 3]
```diff
@@ -638,6 +636,23 @@ gboolean* last_seen(const int* arr, int len) {
...
+    for (int i = len - 1; i >= 0; i--) {
+        if (!g_hash_table_contains(seen_values, (arr + i))) {
+            int *key = g_new(gint, 1);
+            *key = arr[i];
+            g_hash_table_add(seen_values, key);
+            count++;
+        }
+    }
+
+    g_hash_table_destroy(seen_values);
```
Comment (tomlegkov) [inline marine.c:652]: Is this a memory leak?
According to this https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-destroy you should free the keys yourself or initialize the hash table with `g_hash_table_new_full`.

**Review by tomlegkov — DISMISSED** [review]

**Pull request merged**